### DISCUSSION TITLE
change node features

### DISCUSSION THREAD

**Discussion by masoudMZB** [pyg-team/pytorch_geometric#7684] [General]
I Loaded Mnist dataset with Torchvision 
`
mnist_test = MNIST(root='./data', train=False, download=True)
`
and used Squeezenet to extract features.  squeeze net gave me 1000 features for each image and I used PCA to reduce them to 32. 
Until here I have a list (10000, 32). These are my features. 
I loaded Mnist superpixel from Pytorch geometric 
`superpixels_test = MNISTSuperpixels(root='./data', train=False)`
I checked if the ordering of these 2 datasets is the same and the answer was yes.

So now I want to change each node feature with extracted features. for example, the first graph has 75,1 dimensionality and I want to change it to 75,32  same feature for all nodes of a graph.

When I run this code I see no change:

```
 dataset = MNISTSuperpixels(root='.', train=False)
 for i,graph in enumerate(dataset):
     num_nodes = graph.num_nodes
     print(graph.x.shape)
     graph.x = torch.tensor(features_pca[i], dtype=torch.float).unsqueeze(0).repeat(num_nodes, 1)
     print(graph.x.shape)
  
 dataset[0]
```

The last print says 75,1 

Is there any other solution for this purpose?

**Comment by akihironitta** [comment]
I hope this code snippet helps:
```python
print(dataset.data.x.shape)  # [750000, 1]
dataset.data.x = torch.ones(75000, 32)
print(dataset.data.x.shape)  # [750000, 32]
```

**Reply by rusty1s** [reply]
Yes, this solution works. In general, it is not recommended to modify `dataset.data` though (ideally we should have made that a private attribute a long time ago). Another way to achieve this is through the usage of `pre_transform`.

**Reply by masoudMZB** [reply]
Can you share sample code or links ? @rusty1s

**Reply by mzamini92** [reply]
This code will loop through all the graphs in the dataset and change the node features for each one. The last print statement will now show that the node features have been changed to 75,32.
```
import torch
from torch_geometric.datasets import MNISTSuperpixels

dataset = MNISTSuperpixels(root='.', train=False)
features_pca = ...  # Your PCA features

for i, graph in enumerate(dataset):
    num_nodes = graph.num_nodes
    graph.x = torch.tensor(features_pca[i], dtype=torch.float).unsqueeze(0).repeat(num_nodes, 1)

print(dataset[0].x.shape)
```
The pre_transform argument in PyTorch Geometric is used to apply a transformation to the data before it is loaded into the dataset. This can be used to change the node features, the edge features, or the graph structure. Here is an example of a pre_transform function that can be used to change the node features:
```
def change_node_features(graph):
    num_nodes = graph.num_nodes
    graph.x = torch.tensor(features_pca[i], dtype=torch.float).unsqueeze(0).repeat(num_nodes, 1)
    return graph
```
To use the pre_transform argument, you would need to pass the change_node_features function to the pre_transform argument when you create the dataset. For example:
```
dataset = MNISTSuperpixels(root='.', train=False, pre_transform=change_node_features)
```
This code will create a dataset with the modified node features.

**Reply by masoudMZB** [reply]
@mzamini92 did you test the code and node features? because I can not get same result and node features still are same.
can you share a google colab?

**Reply by rusty1s** [reply]
This would be something like
```python
class MyTransform:
    def __init__(self, pca):
        self.index = 0
        self.pca = pca
        
    def __call__(self, data):
        data.x = self.pca[self.index]
        self.index += 1
        return data
        
data = MyDataset(..., pre_transform=MyTransform(pca))
